### DISCUSSION TITLE
How to get information about hook points

### DISCUSSION THREAD

**Discussion by try-agaaain** [bpftrace/bpftrace#2706] [Q&A] [answered]
There are over 100,000 hook points listed by `bpftrace -l`. If I want to obtain the description of a specific hook point, where can I find help?

**Comment by fbs** [comment, marked as answer]
The man page has a high level description of each type. https://github.com/iovisor/bpftrace/blob/master/man/adoc/bpftrace.adoc#probes

Most probes hook into the kernel and in most cases the only documentation is the linux kernel source. Generally tracepoints are better documented than kprobe/kfuncs as tracepoints have to be defined in code so they're usually useful for something.

**Comment by try-agaaain** [comment]
Thanks a lot. Can you help me with another question if possible? I wrote a C program named test1, which outputs "hello, world" and the process ID of the program every 0.1 seconds. The main code is as follows:

```c
    while (1) {
        int fd = open("test.txt", O_CREAT | O_WRONLY | O_APPEND, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
        if (fd == -1) {
            perror("Error opening file");
            return 1;
        }

        pid_t pid = getpid();
        printf("Process ID: %d\n", pid);
    
        const char* msg = "hello, world\n";
        write(fd, msg, strlen(msg));
        write(STDOUT_FILENO, msg, strlen(msg));
        usleep(100000); // sleep for 0.1 seconds
        close(fd);
    }
```

The output is "hello, world\nProcess ID: 289352", and the process PID is 289352.

I want to use bpftrace to monitor the open calls in the program. So I checked with `ldd ./test1` and found that the program uses `/lib/x86_64-linux-gnu/libc.so.6`. The mount points related to open in this shared library are as follows:

```shell
$ sudo bpftrace -l 'uprobe:/lib/x86_64-linux-gnu/libc.so.6:open*'
uprobe:/lib/x86_64-linux-gnu/libc.so.6:open
uprobe:/lib/x86_64-linux-gnu/libc.so.6:open64
...
```

Based on this information, I wrote the following command to monitor the open calls of the running test1 program: `sudo bpftrace -e 'uprobe:/lib/x86_64-linux-gnu/libc.so.6:open /pid==289352/ {printf("use open func: %s %d\n", comm, pid);}'`, but there is no output.

Later, I removed the condition `/pid==289352/`, and in the results obtained, the pid of test1 is `1011838`, as follows:

```
...
use open func: node 919420
use open func: test1 1011838
use open func: python 65514
...
```

However, when I use `top -p 1011838`,  I find this process does not exist. I am very confused about why `sudo bpftrace -e 'uprobe:/lib/x86_64-linux-gnu/libc.so.6:open /pid==289352/ {printf("use open func: %s %d\n", comm, pid);}'` does not work correctly, while `sudo bpftrace -e 'uprobe:/lib/x86_64-linux-gnu/libc.so.6:open /pid==1011838/ {printf("use open func: %s %d\n", comm, pid);}'` works correctly.

**Reply by fbs** [reply]
the pids are strangly high, are you tracing directly on linux or is there something in between?

**Reply by try-agaaain** [reply]
Yes, it's strange, I run it in WSL2.

Here are info about my wsl2 environment
```
$ uname -a
Linux team317 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

When I use `/comm=="test1"/` as a filtering condition, I found that only one process with pid 1011838 matches. It should correspond to the process with pid 289352. 

This seems to indicate that in my experiment, the process PIDs obtained by bpftrace are not entirely consistent with the actual PIDs in the system.

**Reply by try-agaaain** [reply]
I repeated the experiment, and this time I found that the PID in the system is `486821`, while the PID captured by bpftrace is `1209544.` The difference between the two is `1209544 - 486821 = 722723`. 

In the previous experiment, the difference between the two was `1011838 - 289352 = 722486`. 

Is it possible that this discrepancy is due to inconsistent data types?

**Reply by fbs** [reply]
Im not familiar with WSL but I expect bpftrace to run in a different namespace. So the pid your application reports is not the pid bpftrace sees. That the offset matches is probably do to the way uids are mapped

**Reply by try-agaaain** [reply]
If I know the PID of a process in a namespace, is there a way to obtain the corresponding PID as seen by BPF? 

Alternatively, if the PID seen by BPF is known, is there a way to obtain the PID in the namespace?
